File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    # 친구들의 이름을 인덱스로 변환한 딕셔너리
    friend_to_index = {friend: index for index, friend in enumerate(friends)}
    n = len(friends)

    # 친구들끼리 주고 받은 선물을 저장할 2차원 리스트
    gift_counts = [[0]*n for _ in range(n)]

    # 2차원 리스트에 주고 받은 선물을 저장
    for gift in gifts:
        giver, receiver = gift.split()
        i, j = friend_to_index[giver], friend_to_index[receiver]
        gift_counts[i][j] += 1

    # 다음 달에 줄 선물의 개수를 저장할 리스트
    gifts_next_month = [0]*n
    
    # 다음 달에 줄 선물의 개수를 계산 
    # 1. i가 j에게 준 선물이 j가 i에게 준 선물보다 많을 때
    # 2. i가 j에게 준 선물과 j가 i에게 준 선물이 같을 때
    # 3. i가 받은 선물의 합이 j가 받은 선물의 합보다 많을 때
    for i in range(n):
        for j in range(n):
            if i != j: # 자기 자신에게 선물을 주는 경우를 제외하고
                if (gift_counts[i][j] > gift_counts[j][i] or # i가 j에게 준 선물이 j가 i에게 준 선물보다 많을 때
                        (gift_counts[i][j] == gift_counts[j][i] and # i가 j에게 준 선물과 j가 i에게 준 선물이 같을 때
                         sum(gift_counts[i]) - sum(row[i] for row in gift_counts) > sum(gift_counts[j]) - sum(row[j] for row in gift_counts))): # i가 받은 선물이 j가 받은 선물보다 많을 때
                    gifts_next_month[i] += 1 # 다음 달에 i는 선물을 받음
                    
    return max(gifts_next_month)

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.kt
class Solution {
    fun solution(friends: Array<String>, gifts: Array<String>): Int {
        val n = friends.size

        // 친구들의 이름을 인덱스로 변환한 맵을 생성
        val friendToIndex = friends.withIndex().associate { it.value to it.index }

        // 친구들끼리 주고 받은 선물을 저장할 2차원 리스트 초기화
        val giftCounts = Array(n) { IntArray(n) }

        // 주고 받은 선물 기록
        for (gift in gifts) {
            val (giver, receiver) = gift.split(" ")
            val i = friendToIndex[giver]!!
            val j = friendToIndex[receiver]!!
            giftCounts[i][j]++
        }

        // 다음 달에 줄 선물의 개수를 저장할 배열
        val giftsNextMonth = IntArray(n)

        // 다음 달에 줄 선물의 개수 계산
        for (i in 0 until n) {
            for (j in 0 until n) {
                if (i != j && giftCounts[i][j] > giftCounts[j][i]) {
                    giftsNextMonth[i]++
                } else if (i != j && giftCounts[i][j] == giftCounts[j][i]) {
                    val giveCountI = giftCounts[i].sum() - giftCounts.sumBy { it[i] }
                    val giveCountJ = giftCounts[j].sum() - giftCounts.sumBy { it[j] }
                    if (giveCountI > giveCountJ) {
                        giftsNextMonth[i]++
                    }
                }
            }
        }

        return giftsNextMonth.maxOrNull() ?: 0
    }
}
